### ISSUE TITLE
wkhtmltopdf process exited with code 126

### ISSUE THREAD

**Issue by Luke-1988** [zeplin/zeplin-html-to-pdf#15]
Hi guys,

I am getting this error when deployed on Lambda:

{
  "errorMessage": "[ServerError] Internal server error\nError: wkhtmltopdf process exited with code 126\n    at ChildProcess.proc.on.on.code (/var/task/utils/wkhtmltopdf.js:12:24)\n    at emitTwo (events.js:106:13)\n    at ChildProcess.emit (events.js:191:7)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)"
}

Do you have any hint?
Thanks!

**Comment by bdegley4789** [comment]
{
    "errorMessage": "[ServerError] Internal server error\nError: wkhtmltopdf process exited with code 127\n    at ChildProcess.proc.on.on.code (/var/task/app/utils/wkhtmltopdf.js:12:24)\n    at emitTwo (events.js:126:13)\n    at ChildProcess.emit (events.js:214:7)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)"
}

**Comment by bdegley4789** [comment]
@Luke-1988 did you ever resolve this?

**Comment by Luke-1988** [comment]
No success here. 

I found another solution: https://github.com/jblotus/aws-lambda-wkhtmltopdf
However, this solution works only when I upload to AWS the zip file referenced in repository: https://github.com/lubos/aws-lambda-wkhtmltopdf/releases/download/0.1/wkhtmltopdf.zip

When I clone repository and zip files on my own, it does not work.

**Comment by rachnakela** [comment]
Did you find the solution?

**Comment by rachnakela** [comment]
I am getting following exception - 
{
  "errorType": "string",
  "errorMessage": "[ServerError] Internal server error\nError: wkhtmltopdf process exited with code 127\n    at ChildProcess.proc.on.on.code (/var/task/utils/wkhtmltopdf.js:12:24)\n    at ChildProcess.emit (events.js:198:13)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)",
  "trace": []
}

**Comment by virendra656** [comment]
> No success here.
> 
> I found another solution: https://github.com/jblotus/aws-lambda-wkhtmltopdf
> However, this solution works only when I upload to AWS the zip file referenced in repository: https://github.com/lubos/aws-lambda-wkhtmltopdf/releases/download/0.1/wkhtmltopdf.zip
> 
> When I clone repository and zip files on my own, it does not work.

You need to set permission for the binary

**Comment by dorukgunes** [comment]
This issue should have been fixed with https://github.com/zeplin/zeplin-html-to-pdf/pull/19, could you please try again with a runtime >=nodejs6.10, feel free to open again if you encounter the same problem.

**Comment by hectorgrecco** [comment]
I'm having the same issue... Runtime node.js 12.x

**Comment by 0rbster** [comment]
Same problem for me with runtime node.js 12.x. Is there any fix for this?

**Comment by Bastidias** [comment]
Hi, 

I was having issues with exited 127 but managed to get it working. 

**tldr:**
- Confirm location of the wkhtmltopdf binary
- Add execute permissions to the wkhtmltopdf


**Error**
```
2020-05-20T12:00:44.390Z	6fad6f08-9c96-4c53-9554-4603b7fb6d95	WARN	Error: wkhtmltopdf process exited with code 127
    at ChildProcess.<anonymous> (/var/task/createpdf.js:452:24)
    at ChildProcess.emit (events.js:310:20)
    at ChildProcess.EventEmitter.emit (domain.js:482:12)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
```

**Confirm the location of the binary:**

On my deployed function, I editing ./utils/wkhtmltopdf.js and changed it to the following:
``` 
const { spawn, exec } = require("child_process");
module.exports = function (html, options = []) {
  let cmd = 'ls -lha ./lib'
  const ls = exec(cmd, function(error, stdout, stderr) {
        if (error) {
          console.log(error.stack);
          console.log(error.code);
          reject(error);
        } else {
          resolve(stdout);
        }
        console.log('stdlout\n' + stdout);
        console.log(stderr);
      });
    
      ls.on('exit', function(code) {
        console.log('exited with' + code);
      })
};
```

I tried different iterations of ``` let cmd = ``` such has
``` let cmd = 'ls -lha' ```
``` let cmd = 'ls -lha ./lib' ```

My first attempts confirmed that I was not executing from the correct location. 
When I finally found the right folder path, I saw the issue with permissions:
```
-rw-r--r-- 1 root root  38M Jan  1  1980 wkhtmltopdf  <----- No execute privs!
```

To fix this, I used the ```webpack-permissions-plugin``` and ```copy-webpack-plugin``` to add the ./lib and ./font folder to my function zip and change the privs. 

**Folder Structure**:
```
./src/ -- the rest of the js code
./src/utils/wkhtmlpdf.js <-- from the repo

./serverless.yml
./webpack.config.js
./libs/lib/ <-- from repo
./libs/font/ <-- from repo
./.webpack <-- webpack build folder
```

**webpack.config.js**
```
const path = require("path");
const nodeExternals = require("webpack-node-externals");
const slsw = require("serverless-webpack");
const CopyPlugin = require("copy-webpack-plugin");
const PermissionsOutputPlugin = require('webpack-permissions-plugin');

module.exports = {
  entry: slsw.lib.entries,
  target: "node",
  externals: [nodeExternals()],
  output: {
    libraryTarget: "commonjs",
    path: path.join(__dirname, ".webpack"),
    filename: "[name].js",
  },
  plugins: [
    new CopyPlugin({
      patterns: [{ 
        from: path.resolve(__dirname, "libs"),
        to: path.resolve(__dirname, '.webpack/MY-LAMBDA-FUNCTION')
      }],
    }),
    new PermissionsOutputPlugin({
      buildFiles: [
          path.resolve(
            __dirname,
              '.webpack',
              'MY-LAMBDA-FUNCTION',
              'lib',
              'wkhtmltopdf'
          ),
      ],
  }),
  ],
  module: {
    rules: [
      {
        test: /\.js$/,
        loader: "babel-loader",
        include: __dirname,
        exclude: /node_modules/,
      },
    ],
  },
};

```
After with the permissions plugin:
```
-rwxr-xr-x 1 root root  38M Jan  1  1980 wkhtmltopdf  <----- Yay, we can execute!
```

At this point, I reverted back the original wkhtmltopdf.js code and was able to generate a PDF. 

**Side note:**
add ```console.log(process.env)``` to your debugging to make sure that you have your font and path envs set correctly.
